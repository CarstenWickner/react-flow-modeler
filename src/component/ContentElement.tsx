import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
