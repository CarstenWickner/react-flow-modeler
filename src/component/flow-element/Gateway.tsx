import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const Gateway: React.FunctionComponent<{ type: "converging" } | { type: "diverging"; children?: React.ReactChild }> = ({ type, children }) => (
    <>
        <div className="arrow" />
        <div className={`flow-element gateway-element ${type}`} />
        {type === "diverging" && <HorizontalStroke>{children}</HorizontalStroke>}
    </>
);
