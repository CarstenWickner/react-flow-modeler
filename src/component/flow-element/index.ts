export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
