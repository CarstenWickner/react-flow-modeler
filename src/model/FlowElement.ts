export interface FlowElementReference {
    getId: () => string;
    getPrecedingElements: () => Array<FlowElementReference>;
    getFollowingElements: () => Array<FlowElementReference>;
}

export class FlowElement implements FlowElementReference {
    private id: string;
    private columnIndex: number;
    private rowCount: number;
    private precedingElements: Array<FlowElement>;
    private followingElements: Array<FlowElement>;

    constructor(id: string) {
        this.id = id;
        this.precedingElements = [];
        this.followingElements = [];
    }

    getId(): string {
        return this.id;
    }

    getColumnIndex(): number {
        return this.columnIndex;
    }

    setColumnIndex(columnIndex: number): void {
        this.columnIndex = columnIndex;
    }

    getRowCount(): number {
        return this.rowCount;
    }

    setRowCount(rowCount: number): void {
        this.rowCount = rowCount;
    }

    getPrecedingElements(): Array<FlowElement> {
        return this.precedingElements;
    }

    addPrecedingElement(element: FlowElement): void {
        this.precedingElements.push(element);
    }

    getFollowingElements(): Array<FlowElement> {
        return this.followingElements;
    }

    addFollowingElement(element: FlowElement): void {
        this.followingElements.push(element);
    }
}
