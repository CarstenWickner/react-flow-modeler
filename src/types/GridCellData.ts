export interface GridCellCoordinates {
    colStartIndex: number;
    colEndIndex?: number;
    rowStartIndex: number;
    rowEndIndex?: number;
}

export interface SimpleElement extends GridCellCoordinates {
    type: "start" | "end" | "stroke-extension";
}

/*
 * contextual information concerning a flow element
 */
export interface ContentCellData extends GridCellCoordinates {
    type: "content";
    elementId: string;
    data?: { [key: string]: unknown };
}
export interface GatewayDivergingCellData extends GridCellCoordinates {
    type: "gateway-diverging";
    gatewayId: string;
    data?: { [key: string]: unknown };
}
export interface GatewayToElementConnectorCellData extends GridCellCoordinates {
    type: "gateway-to-element";
    gatewayId: string;
    elementId?: string;
    data?: { [key: string]: unknown };
    connectionType: "first" | "middle" | "last";
}
export interface ElementToGatewayConnectorCellData extends GridCellCoordinates {
    type: "element-to-gateway";
    elementId: string;
    connectionType: "first" | "middle" | "last";
}
export interface GatewayConvergingCellData extends GridCellCoordinates {
    type: "gateway-converging";
}

export type GridCellData =
    | SimpleElement
    | ContentCellData
    | GatewayDivergingCellData
    | GatewayToElementConnectorCellData
    | GatewayConvergingCellData
    | ElementToGatewayConnectorCellData;
