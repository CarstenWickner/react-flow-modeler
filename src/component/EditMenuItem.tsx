import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
