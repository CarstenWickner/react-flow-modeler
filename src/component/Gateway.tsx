import * as React from "react";

import { FlowElementWrapper } from "./FlowElementWrapper";
import { HorizontalStroke } from "./HorizontalStroke";

import { ConvergingGatewayNode, DivergingGatewayNode, ElementType, DivergingGatewayBranch } from "../types/ModelElement";
import { onLinkDropCallback } from "../types/EditAction";

const returnNull = (): null => null;

export class Gateway extends React.Component<{
    gateway: DivergingGatewayNode | ConvergingGatewayNode;
    onSelect: (reference: DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode) => void;
    editMenu: (() => React.ReactNode) | undefined;
    onLinkDrop: onLinkDropCallback | undefined;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, gateway } = this.props;
        onSelect(gateway);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { gateway, editMenu, onLinkDrop: onDrop, onSelect, children } = this.props;
        const cssType = gateway.type == ElementType.DivergingGatewayNode ? "diverging" : "converging";
        return (
            <>
                <FlowElementWrapper
                    elementTypeClassName={`gateway-element ${cssType}`}
                    referenceElement={gateway}
                    editMenu={editMenu}
                    onLinkDrop={onDrop}
                    onClick={this.onClick}
                />
                {gateway.type === ElementType.ConvergingGatewayNode && <div className="stroke-horizontal optional" />}
                {gateway.type === ElementType.DivergingGatewayNode && (
                    // if this gateway is selected, provide a function as editMenu to indicate that, but let the function not render a second menu
                    <HorizontalStroke referenceElement={gateway} editMenu={children && editMenu ? returnNull : undefined} onSelect={onSelect}>
                        {children}
                    </HorizontalStroke>
                )}
            </>
        );
    }
}
