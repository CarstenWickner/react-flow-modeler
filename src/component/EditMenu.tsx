import * as React from "react";

import { EditMenuItem } from "./EditMenuItem";
import { FlowElementReference } from "../model/FlowElement";
import { addContentElement, addDivergingGateway } from "../model/action/addElement";
import { addBranch } from "../model/action/addBranch";
import { isChangeNextElementAllowed } from "../model/action/changeNextElement";
import { removeElement, isRemoveElementAllowed } from "../model/action/removeElement";

import { SelectableElementType, EditActionResult, DraggableType } from "../types/EditAction";
import { FlowModelerProps, MenuOptions } from "../types/FlowModelerProps";
import { ElementType } from "../types/GridCellData";

const onClickStopPropagation = (event: React.MouseEvent): void => event.stopPropagation();

export class EditMenu extends React.Component<{
    targetType: SelectableElementType;
    referenceElement?: FlowElementReference;
    branchIndex?: number;
    menuOptions?: FlowModelerProps["options"]["editActions"];
    onChange: (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult) => void;
}> {
    isNextElementReferencedByOthers = (): boolean => {
        const { referenceElement, branchIndex } = this.props;
        if (!referenceElement) {
            return false;
        }
        return referenceElement.getFollowingElements()[branchIndex || 0].getPrecedingElements().length > 1;
    };

    renderMenuItem(
        options: MenuOptions,
        defaultClassName: string,
        onClick?: (event: React.MouseEvent) => void,
        dragType?: DraggableType
    ): React.ReactNode {
        const { targetType, referenceElement, branchIndex } = this.props;
        if (options && options.isActionAllowed && !options.isActionAllowed(targetType, referenceElement, branchIndex)) {
            return null;
        }
        const dragItem =
            dragType === DraggableType.LINK &&
            (targetType === ElementType.Content || targetType == ElementType.ConnectGatewayToElement) &&
            this.isNextElementReferencedByOthers()
                ? { type: dragType, originType: targetType, originElement: referenceElement, originBranchIndex: branchIndex }
                : undefined;
        return <EditMenuItem key={defaultClassName} options={options} defaultClassName={defaultClassName} onClick={onClick} dragItem={dragItem} />;
    }

    onAddContentElementClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addContentElement(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddContentElementItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addFollowingContentElement : undefined, "add-content", this.onAddContentElementClick);
    }

    onAddDivergingGatewayClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addDivergingGateway(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddDivergingGatewayItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(
            menuOptions ? menuOptions.addFollowingDivergingGateway : undefined,
            "add-gateway",
            this.onAddDivergingGatewayClick
        );
    }

    onAddDivergingBranchClick = (): void => {
        const { targetType, onChange, referenceElement } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            onChange((originalFlow) => addBranch(originalFlow, {}, referenceElement));
        }
    };

    renderAddDivergingBranchItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addDivergingBranch : undefined, "add-branch", this.onAddDivergingBranchClick);
    }

    renderChangeNextElementItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isChangeNextElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.changeNextElement : undefined, "change-next", undefined, DraggableType.LINK);
        }
        return null;
    }

    onRemoveClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType === ElementType.Content || targetType === ElementType.ConnectGatewayToElement) {
            onChange((originalFlow) => removeElement(originalFlow, targetType, referenceElement, branchIndex));
        }
    };

    renderRemoveItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isRemoveElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.removeElement : undefined, "remove", this.onRemoveClick);
        }
        return null;
    }

    render(): React.ReactNode {
        const menuItems = [
            this.renderAddContentElementItem(),
            this.renderAddDivergingGatewayItem(),
            this.renderAddDivergingBranchItem(),
            this.renderChangeNextElementItem(),
            this.renderRemoveItem()
        ];
        if (menuItems.some((item) => item)) {
            return (
                <div className="menu" onClick={onClickStopPropagation}>
                    {menuItems}
                </div>
            );
        }
        return null;
    }
}
