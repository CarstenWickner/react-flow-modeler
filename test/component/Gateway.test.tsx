import * as React from "react";
import { mount, shallow } from "enzyme";

import { Gateway } from "../../src/component/Gateway";
import { ElementType } from "../../src/types/GridCellData";
import { FlowElementReference } from "../../src/model/FlowElement";

const mockFlowElementReference = (id: string): FlowElementReference => ({
    getId: (): string => id,
    getPrecedingElements: (): Array<FlowElementReference> => [],
    getFollowingElements: (): Array<FlowElementReference> => []
});

describe("renders correctly", () => {
    const onSelect = (): void => {};
    describe("as converging gateway", () => {
        it("with minimal props", () => {
            const component = shallow(
                <Gateway
                    type={ElementType.GatewayConverging}
                    followingElement={mockFlowElementReference("following-element-id")}
                    editMenu={(): React.ReactNode => <div className="edit-menu-placeholder" />}
                    onLinkDrop={undefined}
                    onSelect={onSelect}
                />
            );
            expect(component).toMatchSnapshot();
        });
        it("when not selected", () => {
            const component = mount(
                <Gateway
                    type={ElementType.GatewayConverging}
                    followingElement={mockFlowElementReference("following-element-id")}
                    editMenu={undefined}
                    onLinkDrop={undefined}
                    onSelect={onSelect}
                />
            );
            expect(component.find(".gateway-element").hasClass("selected")).toBe(false);
            expect(component.find(".edit-menu-placeholder").exists()).toBe(false);
        });
    });
    describe("as diverging gateway", () => {
        it("with minimal props", () => {
            const component = shallow(
                <Gateway
                    type={ElementType.GatewayDiverging}
                    gateway={mockFlowElementReference("gateway-id")}
                    editMenu={(): React.ReactNode => <div className="edit-menu-placeholder" />}
                    onLinkDrop={undefined}
                    onSelect={onSelect}
                >
                    {"text"}
                </Gateway>
            );
            expect(component).toMatchSnapshot();
        });
        it("when selected without children", () => {
            const component = mount(
                <Gateway
                    type={ElementType.GatewayDiverging}
                    gateway={mockFlowElementReference("gateway-id")}
                    editMenu={(): React.ReactNode => <div className="edit-menu-placeholder" />}
                    onLinkDrop={undefined}
                    onSelect={onSelect}
                />
            );
            expect(component.find(".gateway-element").hasClass("selected")).toBe(true);
            expect(component.find(".edit-menu-placeholder").exists()).toBe(true);
            expect(component.find("HorizontalStroke").prop("editMenu")).toBe(undefined);
        });
        it("when not selected", () => {
            const component = mount(
                <Gateway
                    type={ElementType.GatewayDiverging}
                    gateway={mockFlowElementReference("gateway-id")}
                    editMenu={undefined}
                    onLinkDrop={undefined}
                    onSelect={onSelect}
                >
                    {"text"}
                </Gateway>
            );
            expect(component.find(".gateway-element").hasClass("selected")).toBe(false);
            expect(component.find(".edit-menu-placeholder").exists()).toBe(false);
            expect(component.find("HorizontalStroke").prop("editMenu")).toBe(undefined);
        });
    });
});
describe("calls onSelect", () => {
    const onSelect = jest.fn(() => {});
    const event = ({ stopPropagation: jest.fn(() => {}) } as unknown) as React.MouseEvent;
    it("on click event for diverging gateway", () => {
        const component = mount(
            <Gateway
                type={ElementType.GatewayDiverging}
                editMenu={undefined}
                gateway={mockFlowElementReference("gateway-id")}
                onLinkDrop={undefined}
                onSelect={onSelect}
            >
                {"text"}
            </Gateway>
        );
        component.find(".gateway-element").prop("onClick")(event);
        expect(onSelect.mock.calls).toHaveLength(1);
        expect(onSelect.mock.calls[0]).toHaveLength(1);
        expect(onSelect.mock.calls[0][0]).toEqual("gateway-id");
    });
    it("on click event for converging gateway", () => {
        const component = mount(
            <Gateway
                type={ElementType.GatewayConverging}
                editMenu={undefined}
                followingElement={mockFlowElementReference("following-element-id")}
                onLinkDrop={undefined}
                onSelect={onSelect}
            />
        );
        component.find(".gateway-element").prop("onClick")(event);
        expect(onSelect.mock.calls).toHaveLength(1);
        expect(onSelect.mock.calls[0]).toHaveLength(1);
        expect(onSelect.mock.calls[0][0]).toEqual("following-element-id");
    });
});
