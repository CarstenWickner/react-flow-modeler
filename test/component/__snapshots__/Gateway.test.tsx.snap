// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly as converging gateway with minimal props 1`] = `
<Fragment>
  <Component
    editMenu={[Function]}
    elementTypeClassName="gateway-element converging"
    onClick={[Function]}
    referenceElement={FlowElementReference { "id": "following-element-id" }}
  />
  <HorizontalStroke
    optional={true}
  />
</Fragment>
`;

exports[`renders correctly as diverging gateway with minimal props 1`] = `
<Fragment>
  <Component
    editMenu={[Function]}
    elementTypeClassName="gateway-element diverging"
    onClick={[Function]}
    referenceElement={FlowElementReference { "id": "gateway-id" }}
  />
  <HorizontalStroke
    editMenu={[Function]}
    gatewayId="gateway-id"
    onSelect={[Function]}
  >
    text
  </HorizontalStroke>
</Fragment>
`;
